Fourier-Chebyshev Fit to `B` for particle tracing

YigitElma opened the pull request:
- Adds a new convenience class `FourierChebyshevField` which has a couple of methods to be compatible with optimization,
  - `init` : only resolution info
  - `build` : creates grid and transforms
  - `fit` : given params and profiles (these change during optimization), fits a new Fourier-Chebyshev series. 
  - `evaluate` : Evaluate the series at requested point.

Example usage will be,
```python
interpolater = FourierChebyshevField(L=eq.L_grid, M=eq.M_grid, N=eq.N_grid)
interpolater.build(eq)
data_fit = interpolater.fit(eq.params_dict, {"iota": eq.iota})

rtz, _, aux = _trace_particles(
    field=interpolator,  # normally you pass eq here
    y0=x0,
    model=model,
    model_args=args,
    ts=ts,
    params=None,   # normally you pass eq.params_dict here
    max_steps=int((ts[-1] - ts[0]) / 1e-8),
    min_step_size=1e-8,
    stepsize_controller=PIDController(rtol=1e-4, atol=1e-4, dtmin=1e-8),
    saveat=SaveAt(ts=ts),
    solver=Tsit5(),
    adjoint=RecursiveCheckpointAdjoint(),
    event=event,
    chunk_size=None,
    options={},    # this passes extra stuff to field computation, not needed here
    throw=False,
    return_aux=True,
)
```

Todo:
- [ ] `trace_particles` should initialize `FourierChebyshevField` and by default use it. Add a flag to use Equilibrium field
- [ ] `plot_particle_trajectories` etc should also be able to use the interpolated field.
- [ ] Look at the spectral convergence (ideally lower res for basis vectors might be good, but current code is simpler and with batching might not be inefficient)
- [ ] Consider adding spline version too?

f0uriest committed:
Merge branch 'dp/field-line-integrate-diffrax' into rc/coil_tracing

f0uriest committed:
Refactor to use diffrax

YigitElma committed:
Merge branch 'master' into rc/coil_tracing

f0uriest committed:
Add option for full orbit

joaopedrobiu6 committed:
Merge branch 'PlasmaControl:master' into jb/DPS

joaopedrobiu6 committed:
Merge branch 'PlasmaControl:master' into jb/DPS

joaopedrobiu6 committed:
Merge branch 'PlasmaControl:master' into jb/DPS

dpanici committed:
Merge branch 'master' into jb/DPS

joaopedrobiu6 committed:
Merge DPS from fork to jb/DPS branch on DESC repository (#1625)

joaopedrobiu6 committed:
Now it automatically chooses if it solves 1 particle or more than 1 particle based on the dimension of the initial conditions. `particle_tracer.py` and `optimization.py` are adapted to this change.
Optimization doesn't work yet

joaopedrobiu6 committed:
Single or multiparticle adaptation (#1626)

Automatically chooses single or multiparticle based on the dimension of
the initial conditions. `particle_tracer.py` and `optimization.py` are
adapted to this change.

Optimization doesn't work yet

dpanici committed:
update script

f0uriest committed:
Formatting

f0uriest committed:
Merge branch 'master' into rc/coil_tracing

f0uriest committed:
Merge branch 'master' into jb/DPS

f0uriest committed:
Merge branch 'rc/coil_tracing' into rc/particles

f0uriest committed:
Add basic sketch of particle tracing module

f0uriest committed:
Add better implementation of coulomb log and slowing down time

f0uriest committed:
Add initializer for real space

f0uriest committed:
Add args to trajectory models

f0uriest committed:
Allow models to work in either frame

f0uriest committed:
Clean up initializers to work with args in either frame

f0uriest committed:
Add diffrax stuff to tracing, add type hints

f0uriest committed:
Remove old stuff for tracing particles in coils

f0uriest committed:
Add tracing to __init__

f0uriest committed:
Add surface initializer, initializers now work i think

f0uriest committed:
Fix class structure for model/diffrax terms

dpanici committed:
Merge branch 'master' into rc/particles

YigitElma committed:
minor clean-up, move test scripts to private repo

YigitElma committed:
remove DS_store

YigitElma committed:
rename files as particles

YigitElma committed:
minor clean-up, update docs, use scipy constants

YigitElma committed:
update docs

YigitElma committed:
add bounds terminator

YigitElma committed:
simplify some equations

YigitElma committed:
simplify more equations, add some comments on possible mistakes

YigitElma committed:
fix typo

YigitElma committed:
rename event

YigitElma committed:
move back to using ODETerm instead of custom term

YigitElma committed:
move integration functions out of model class, make the function vmapped with expected array shapes, needs clean up

YigitElma committed:
clean up, move the functions back to model classes, need to figure out static attributes

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
remove compute functions, only use trace_particles

YigitElma committed:
Merge branch 'rc/particles' of github.com:PlasmaControl/DESC into rc/particles

YigitElma committed:
remove the static attrs

YigitElma committed:
Merge remote-tracking branch 'origin/rc/pytrees' into rc/particles

YigitElma committed:
add static fields properly with equinox

YigitElma committed:
add plotting function, add one example probably will be changed later

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
remove the objective class, it had redundant code for integration, we can add a new objective later

YigitElma committed:
remove the example figure, fix some docs

YigitElma committed:
move common logic to super class

YigitElma committed:
clean up

YigitElma committed:
fix bugs

YigitElma committed:
Merge branch 'rc/pytrees' into rc/particles

YigitElma committed:
remove print

YigitElma committed:
add better warning for flux/lab particles, convert coordinates to proper frames

YigitElma committed:
update changelog

YigitElma committed:
use equinox errors

YigitElma committed:
clean up random indices genrator

YigitElma committed:
Merge remote-tracking branch 'origin/rc/pytrees' into rc/particles

YigitElma committed:
add initial tests

YigitElma committed:
add safediv

YigitElma committed:
add end-point to the plot, fix adaptive stepsize controller for Euler case, use better default max_steps, make arrays 1d

YigitElma committed:
apply euler exception to field_line_integrate, rename maxstep to max_steps for consistency with diffrax, fix example

YigitElma committed:
Update desc/particles.py

Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>

YigitElma committed:
Update desc/magnetic_fields/_core.py

Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>

YigitElma committed:
add equations to docs, add refs, address Dario's comments, fix typos

YigitElma committed:
Merge remote-tracking branch 'origin/rc/pytrees' into rc/particles

YigitElma committed:
fixes for docs

YigitElma committed:
make definition of mu mathc the convention, fix equations, add test

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
fix equations, add 1 more test

YigitElma committed:
add new test, divide field velocity by R for angular speed, fix typos

dpanici committed:
not sure why, but Rdot seems to be 2D and so changed the at,set statement accordingly. works now

dpanici committed:
Merge branch 'master' into rc/particles

dpanici committed:
fix some bugs: shape errors in vac guiding center model vf, and ensuring MagneticFieldFromUser works correctly when its own params_dict is passed in to its compute magnetic field method

dpanici committed:
fix some bugs: shape errors in vac guiding center model vf, and ensuring MagneticFieldFromUser works correctly when its own params_dict is passed in to its compute magnetic field method

dpanici committed:
Merge branch 'rc/particles' of github.com:PlasmaControl/DESC into rc/particles

YigitElma committed:
clean up the fix for phidot

YigitElma committed:
minor updates

YigitElma committed:
add initializer tests, add proper error/warnings, misc clean-up

YigitElma committed:
update docs

YigitElma committed:
fix the smal bug

YigitElma committed:
Update tests/test_particles.py

YigitElma committed:
Update tests/test_particles.py

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

YigitElma committed:
fix docs, remove type hints, add test

YigitElma committed:
fix docs

YigitElma committed:
try to solve the weird doc failure

YigitElma committed:
remove the eq from example code

YigitElma committed:
add a way to pass in iota profile

YigitElma committed:
Merge remote-tracking branch 'origin/master' into rc/particles

YigitElma committed:
use Event instead of deprecated discrete terminating event

YigitElma committed:
update field_line_integrate to use vmap and event

YigitElma committed:
add rho and theta to plot data

YigitElma committed:
pass model args as array, addres some comments

YigitElma committed:
Merge remote-tracking branch 'origin/master' into rc/particles

YigitElma committed:
update api docs

YigitElma committed:
update bounds

YigitElma committed:
small fixes

YigitElma committed:
fix test

YigitElma committed:
set better default to max_steps, update the docs

YigitElma committed:
add missing options

YigitElma committed:
remove chnages from field_line_integrate

YigitElma committed:
update changelog

dependabot[bot] committed:
Update pre-commit requirement from <=4.2.0 to <=4.3.0 in the linting group (#1844)

YigitElma committed:
update docs

YigitElma committed:
clean up curve/surface paraticle initializers

YigitElma committed:
make plotting faster

YigitElma committed:
fix curve/surface initializer when they are outside of the equilibrium, add tests

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
minor fixes, trace back who added that params dict thing

dependabot[bot] committed:
Bump actions/download-artifact from 4 to 5 (#1843)

Bumps
[actions/download-artifact](https://github.com/actions/download-artifact)
from 4 to 5.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/actions/download-artifact/releases">actions/download-artifact's
releases</a>.</em></p>
<blockquote>
<h2>v5.0.0</h2>
<h2>What's Changed</h2>
<ul>
<li>Update README.md by <a
href="https://github.com/nebuk89"><code>@​nebuk89</code></a> in <a
href="https://redirect.github.com/actions/download-artifact/pull/407">actions/download-artifact#407</a></li>
<li>BREAKING fix: inconsistent path behavior for single artifact
downloads by ID by <a
href="https://github.com/GrantBirki"><code>@​GrantBirki</code></a> in <a
href="https://redirect.github.com/actions/download-artifact/pull/416">actions/download-artifact#416</a></li>
</ul>
<h2>v5.0.0</h2>
<h3>🚨 Breaking Change</h3>
<p>This release fixes an inconsistency in path behavior for single
artifact downloads by ID. <strong>If you're downloading single artifacts
by ID, the output path may change.</strong></p>
<h4>What Changed</h4>
<p>Previously, <strong>single artifact downloads</strong> behaved
differently depending on how you specified the artifact:</p>
<ul>
<li><strong>By name</strong>: <code>name: my-artifact</code> → extracted
to <code>path/</code> (direct)</li>
<li><strong>By ID</strong>: <code>artifact-ids: 12345</code> → extracted
to <code>path/my-artifact/</code> (nested)</li>
</ul>
<p>Now both methods are consistent:</p>
<ul>
<li><strong>By name</strong>: <code>name: my-artifact</code> → extracted
to <code>path/</code> (unchanged)</li>
<li><strong>By ID</strong>: <code>artifact-ids: 12345</code> → extracted
to <code>path/</code> (fixed - now direct)</li>
</ul>
<h4>Migration Guide</h4>
<h5>✅ No Action Needed If:</h5>
<ul>
<li>You download artifacts by <strong>name</strong></li>
<li>You download <strong>multiple</strong> artifacts by ID</li>
<li>You already use <code>merge-multiple: true</code> as a
workaround</li>
</ul>
<h5>⚠️ Action Required If:</h5>
<p>You download <strong>single artifacts by ID</strong> and your
workflows expect the nested directory structure.</p>
<p><strong>Before v5 (nested structure):</strong></p>
<pre lang="yaml"><code>- uses: actions/download-artifact@v4
  with:
    artifact-ids: 12345
    path: dist
# Files were in: dist/my-artifact/
</code></pre>
<blockquote>
<p>Where <code>my-artifact</code> is the name of the artifact you
previously uploaded</p>
</blockquote>
<p><strong>To maintain old behavior (if needed):</strong></p>
<pre lang="yaml"><code>&lt;/tr&gt;&lt;/table&gt; 
</code></pre>
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/actions/download-artifact/commit/634f93cb2916e3fdff6788551b99b062d0335ce0"><code>634f93c</code></a>
Merge pull request <a
href="https://redirect.github.com/actions/download-artifact/issues/416">#416</a>
from actions/single-artifact-id-download-path</li>
<li><a
href="https://github.com/actions/download-artifact/commit/b19ff4302770b82aa4694b63703b547756dacce6"><code>b19ff43</code></a>
refactor: resolve download path correctly in artifact download tests
(mainly ...</li>
<li><a
href="https://github.com/actions/download-artifact/commit/e262cbee4ab8c473c61c59a81ad8e9dc760e90db"><code>e262cbe</code></a>
bundle dist</li>
<li><a
href="https://github.com/actions/download-artifact/commit/bff23f9308ceb2f06d673043ea6311519be6a87b"><code>bff23f9</code></a>
update docs</li>
<li><a
href="https://github.com/actions/download-artifact/commit/fff8c148a8fdd56aa81fcb019f0b5f6c65700c4d"><code>fff8c14</code></a>
fix download path logic when downloading a single artifact by id</li>
<li><a
href="https://github.com/actions/download-artifact/commit/448e3f862ab3ef47aa50ff917776823c9946035b"><code>448e3f8</code></a>
Merge pull request <a
href="https://redirect.github.com/actions/download-artifact/issues/407">#407</a>
from actions/nebuk89-patch-1</li>
<li><a
href="https://github.com/actions/download-artifact/commit/47225c44b359a5155efdbbbc352041b3e249fb1b"><code>47225c4</code></a>
Update README.md</li>
<li>See full diff in <a
href="https://github.com/actions/download-artifact/compare/v4...v5">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=actions/download-artifact&package-manager=github_actions&previous-version=4&new-version=5)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kaya Unalmis <kayaunalmis@proton.me>

dpanici committed:
Reuse Z transform for L in from_near_axis (#1841)

Reuse the Z transform for lambda in `equilibrium.from_near_axis`; it's
more efficient and requires building fewer large matrices.

YigitElma committed:
use jax.random.choice instead of custom function

YigitElma committed:
add public function trace_particles which takes in initializer and hence makes sure the field, model and initializer are all compatible

YigitElma committed:
move slowdown model to a new PR, this one is getting very big, and we should merge it soon

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
update changelog

YigitElma committed:
update changelog

YigitElma committed:
use cartesian-like coordinate for integration as Rory suggested

YigitElma committed:
add chunk size

YigitElma committed:
move common 3D layout operations to private function, fix docs

YigitElma committed:
fix the saved rtz values in plotting

YigitElma committed:
use compute_fun instead of eq.compute for better performance in particle initializer

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
increase codecov

YigitElma committed:
add rho<1e-6 explanation

YigitElma committed:
Update desc/particles.py

Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>

YigitElma committed:
Merge branch 'master' into rc/particles

unalmis committed:
Fixing stuff from 1529 (#1856)

A new quantity little `gamma_c` was recently added (yesterday) in 1529.
- This quantity was labeled as the fast ion confinement proxy defined by
Dr. Velasco, which is not true as this is the one defined by Nemov.
- There was a factor of $2 /\pi$ factor missing from the new  quantity.
- The wrong coordinate label was assigned to the quantity which might
lead to issues when multiple flux surfaces are requested to be computed
on when using `eq.compute`.

---------

Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>

dpanici committed:
revert action update until issue #1865  is resolved (#1867)

Co-authored-by: Yigit Gunsur Elmacioglu <102380275+YigitElma@users.noreply.github.com>

dependabot[bot] committed:
Bump actions/checkout from 4 to 5 (#1861)

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to
5.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/actions/checkout/releases">actions/checkout's
releases</a>.</em></p>
<blockquote>
<h2>v5.0.0</h2>
<h2>What's Changed</h2>
<ul>
<li>Update actions checkout to use node 24 by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2226">actions/checkout#2226</a></li>
<li>Prepare v5.0.0 release by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2238">actions/checkout#2238</a></li>
</ul>
<h2>⚠️ Minimum Compatible Runner Version</h2>
<p><strong>v2.327.1</strong><br />
<a
href="https://github.com/actions/runner/releases/tag/v2.327.1">Release
Notes</a></p>
<p>Make sure your runner is updated to this version or newer to use this
release.</p>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4...v5.0.0">https://github.com/actions/checkout/compare/v4...v5.0.0</a></p>
<h2>v4.3.0</h2>
<h2>What's Changed</h2>
<ul>
<li>docs: update README.md by <a
href="https://github.com/motss"><code>@​motss</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1971">actions/checkout#1971</a></li>
<li>Add internal repos for checking out multiple repositories by <a
href="https://github.com/mouismail"><code>@​mouismail</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1977">actions/checkout#1977</a></li>
<li>Documentation update - add recommended permissions to Readme by <a
href="https://github.com/benwells"><code>@​benwells</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2043">actions/checkout#2043</a></li>
<li>Adjust positioning of user email note and permissions heading by <a
href="https://github.com/joshmgross"><code>@​joshmgross</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2044">actions/checkout#2044</a></li>
<li>Update README.md by <a
href="https://github.com/nebuk89"><code>@​nebuk89</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2194">actions/checkout#2194</a></li>
<li>Update CODEOWNERS for actions by <a
href="https://github.com/TingluoHuang"><code>@​TingluoHuang</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/2224">actions/checkout#2224</a></li>
<li>Update package dependencies by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2236">actions/checkout#2236</a></li>
<li>Prepare release v4.3.0 by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2237">actions/checkout#2237</a></li>
</ul>
<h2>New Contributors</h2>
<ul>
<li><a href="https://github.com/motss"><code>@​motss</code></a> made
their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/1971">actions/checkout#1971</a></li>
<li><a href="https://github.com/mouismail"><code>@​mouismail</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/1977">actions/checkout#1977</a></li>
<li><a href="https://github.com/benwells"><code>@​benwells</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/2043">actions/checkout#2043</a></li>
<li><a href="https://github.com/nebuk89"><code>@​nebuk89</code></a> made
their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/2194">actions/checkout#2194</a></li>
<li><a href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/2236">actions/checkout#2236</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4...v4.3.0">https://github.com/actions/checkout/compare/v4...v4.3.0</a></p>
<h2>v4.2.2</h2>
<h2>What's Changed</h2>
<ul>
<li><code>url-helper.ts</code> now leverages well-known environment
variables by <a href="https://github.com/jww3"><code>@​jww3</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/1941">actions/checkout#1941</a></li>
<li>Expand unit test coverage for <code>isGhes</code> by <a
href="https://github.com/jww3"><code>@​jww3</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1946">actions/checkout#1946</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4.2.1...v4.2.2">https://github.com/actions/checkout/compare/v4.2.1...v4.2.2</a></p>
<h2>v4.2.1</h2>
<h2>What's Changed</h2>
<ul>
<li>Check out other refs/* by commit if provided, fall back to ref by <a
href="https://github.com/orhantoy"><code>@​orhantoy</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1924">actions/checkout#1924</a></li>
</ul>
<h2>New Contributors</h2>
<ul>
<li><a href="https://github.com/Jcambass"><code>@​Jcambass</code></a>
made their first contribution in <a
href="https://redirect.github.com/actions/checkout/pull/1919">actions/checkout#1919</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/actions/checkout/compare/v4.2.0...v4.2.1">https://github.com/actions/checkout/compare/v4.2.0...v4.2.1</a></p>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a
href="https://github.com/actions/checkout/blob/main/CHANGELOG.md">actions/checkout's
changelog</a>.</em></p>
<blockquote>
<h1>Changelog</h1>
<h2>V5.0.0</h2>
<ul>
<li>Update actions checkout to use node 24 by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2226">actions/checkout#2226</a></li>
</ul>
<h2>V4.3.0</h2>
<ul>
<li>docs: update README.md by <a
href="https://github.com/motss"><code>@​motss</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1971">actions/checkout#1971</a></li>
<li>Add internal repos for checking out multiple repositories by <a
href="https://github.com/mouismail"><code>@​mouismail</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1977">actions/checkout#1977</a></li>
<li>Documentation update - add recommended permissions to Readme by <a
href="https://github.com/benwells"><code>@​benwells</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2043">actions/checkout#2043</a></li>
<li>Adjust positioning of user email note and permissions heading by <a
href="https://github.com/joshmgross"><code>@​joshmgross</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2044">actions/checkout#2044</a></li>
<li>Update README.md by <a
href="https://github.com/nebuk89"><code>@​nebuk89</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2194">actions/checkout#2194</a></li>
<li>Update CODEOWNERS for actions by <a
href="https://github.com/TingluoHuang"><code>@​TingluoHuang</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/2224">actions/checkout#2224</a></li>
<li>Update package dependencies by <a
href="https://github.com/salmanmkc"><code>@​salmanmkc</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/2236">actions/checkout#2236</a></li>
</ul>
<h2>v4.2.2</h2>
<ul>
<li><code>url-helper.ts</code> now leverages well-known environment
variables by <a href="https://github.com/jww3"><code>@​jww3</code></a>
in <a
href="https://redirect.github.com/actions/checkout/pull/1941">actions/checkout#1941</a></li>
<li>Expand unit test coverage for <code>isGhes</code> by <a
href="https://github.com/jww3"><code>@​jww3</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1946">actions/checkout#1946</a></li>
</ul>
<h2>v4.2.1</h2>
<ul>
<li>Check out other refs/* by commit if provided, fall back to ref by <a
href="https://github.com/orhantoy"><code>@​orhantoy</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1924">actions/checkout#1924</a></li>
</ul>
<h2>v4.2.0</h2>
<ul>
<li>Add Ref and Commit outputs by <a
href="https://github.com/lucacome"><code>@​lucacome</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1180">actions/checkout#1180</a></li>
<li>Dependency updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a>- <a
href="https://redirect.github.com/actions/checkout/pull/1777">actions/checkout#1777</a>,
<a
href="https://redirect.github.com/actions/checkout/pull/1872">actions/checkout#1872</a></li>
</ul>
<h2>v4.1.7</h2>
<ul>
<li>Bump the minor-npm-dependencies group across 1 directory with 4
updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1739">actions/checkout#1739</a></li>
<li>Bump actions/checkout from 3 to 4 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1697">actions/checkout#1697</a></li>
<li>Check out other refs/* by commit by <a
href="https://github.com/orhantoy"><code>@​orhantoy</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1774">actions/checkout#1774</a></li>
<li>Pin actions/checkout's own workflows to a known, good, stable
version. by <a href="https://github.com/jww3"><code>@​jww3</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1776">actions/checkout#1776</a></li>
</ul>
<h2>v4.1.6</h2>
<ul>
<li>Check platform to set archive extension appropriately by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1732">actions/checkout#1732</a></li>
</ul>
<h2>v4.1.5</h2>
<ul>
<li>Update NPM dependencies by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1703">actions/checkout#1703</a></li>
<li>Bump github/codeql-action from 2 to 3 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1694">actions/checkout#1694</a></li>
<li>Bump actions/setup-node from 1 to 4 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1696">actions/checkout#1696</a></li>
<li>Bump actions/upload-artifact from 2 to 4 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1695">actions/checkout#1695</a></li>
<li>README: Suggest <code>user.email</code> to be
<code>41898282+github-actions[bot]@users.noreply.github.com</code> by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1707">actions/checkout#1707</a></li>
</ul>
<h2>v4.1.4</h2>
<ul>
<li>Disable <code>extensions.worktreeConfig</code> when disabling
<code>sparse-checkout</code> by <a
href="https://github.com/jww3"><code>@​jww3</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1692">actions/checkout#1692</a></li>
<li>Add dependabot config by <a
href="https://github.com/cory-miller"><code>@​cory-miller</code></a> in
<a
href="https://redirect.github.com/actions/checkout/pull/1688">actions/checkout#1688</a></li>
<li>Bump the minor-actions-dependencies group with 2 updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1693">actions/checkout#1693</a></li>
<li>Bump word-wrap from 1.2.3 to 1.2.5 by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a> in <a
href="https://redirect.github.com/actions/checkout/pull/1643">actions/checkout#1643</a></li>
</ul>
<h2>v4.1.3</h2>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/actions/checkout/commit/08c6903cd8c0fde910a37f88322edcfb5dd907a8"><code>08c6903</code></a>
Prepare v5.0.0 release (<a
href="https://redirect.github.com/actions/checkout/issues/2238">#2238</a>)</li>
<li><a
href="https://github.com/actions/checkout/commit/9f265659d3bb64ab1440b03b12f4d47a24320917"><code>9f26565</code></a>
Update actions checkout to use node 24 (<a
href="https://redirect.github.com/actions/checkout/issues/2226">#2226</a>)</li>
<li>See full diff in <a
href="https://github.com/actions/checkout/compare/v4...v5">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=actions/checkout&package-manager=github_actions&previous-version=4&new-version=5)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Yigit Gunsur Elmacioglu <102380275+YigitElma@users.noreply.github.com>
Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>

dpanici committed:
update changelog with v0.15.0 (#1874)

Co-authored-by: Yigit Gunsur Elmacioglu <102380275+YigitElma@users.noreply.github.com>

unalmis committed:
Use an orthogonal basis for the straight field line equilibrium (#1880)

rahulgaur104 committed:
Adding metric elements and their derivatives in PEST coordinates and other stability-related quantities (#1778)

Aims to resolve #1771

Add higher order derivatives of the position vector in PEST coordinates
and the following metric elements:
$g^{\rho \rho}, g^{\rho \vartheta}. g^{\rho \phi}, g_{\rho \rho},
g_{\rho \vartheta}, g_{\rho \phi}, g_{\vartheta \vartheta}, g_{\phi
\phi}$
and their first order derivatives w.r.t the PEST coordinates $(\rho,
\vartheta, \phi)$.
In addition we add more quantities such as
$J^{\vartheta}, J^{\phi}$
$\mathbf{B} \cdot \mathbf{\nabla} \mathbf{\nabla}\rho$

These quantities are needed for a global MHD stability solver.

---------

Co-authored-by: Dario Panici <37969854+dpanici@users.noreply.github.com>
Co-authored-by: unalmis <kayaunalmis@proton.me>

YigitElma committed:
move default value setters to trace_particles

YigitElma committed:
accept arrays for E, m , q

YigitElma committed:
use proper jax random uniform

YigitElma committed:
make most initializers jitable, only curve to be done by initial guess

YigitElma committed:
maybe make the curve initializer jitable

YigitElma committed:
add jitable tests, allow initializing lab frame by flux inputs through eq kwarg

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
update changelog

YigitElma committed:
add option to skip coordinate mapping for curve and surface initializers

YigitElma committed:
change of variable name

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
minor changes, add chunking to coordinate mapping in plotting

YigitElma committed:
update min step size, an alpha particle moves only couple milimeters in 1e-10 seconds which is unnecessaryly small

dpanici committed:
Merge branch 'master' into rc/particles

YigitElma committed:
doc related fixes

YigitElma committed:
add equinox to requirements explicitly

YigitElma committed:
rename ensure variables

YigitElma committed:
reoder lines

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
refactor inputs as Rory suggested

YigitElma committed:
refactor inputs as Rory suggested part 2

YigitElma committed:
revert the R0 change

YigitElma committed:
add default grid for initializer classes

YigitElma committed:
reduce tolerance, fix typo

YigitElma committed:
add throw argument

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
remove the misleading 'such as'

YigitElma committed:
add aux data, FIX important bug about Grid creation!

YigitElma committed:
add guard against other possible bugs

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
remove the grid fix for now

dpanici committed:
Merge branch 'master' into rc/particles

YigitElma committed:
address Dario's comments

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
fix docs error

YigitElma committed:
fix docs

YigitElma committed:
use paratmetrization instead of object

dpanici committed:
Merge branch 'master' into rc/particles

YigitElma committed:
i don't know what the hell it does but it works! CPU stuff is fine nowpytest -k test_tracing_vacuum_tokamakpytest -k test_tracing_vacuum_tokamak

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Update desc/plotting.py

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
remove redundant comment

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'master' into rc/particles

dpanici committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
add back zernipax function for cpu only, use master one for gpu only. pretty sure other devs will want this in a new PR

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
some fixes

YigitElma committed:
nondiff argument solved the problem

YigitElma committed:
some more fixes

YigitElma committed:
revert zernike changes, it needs special care

YigitElma committed:
add an option to use Fourier Chebyshev interpolation for the magnetic field during particle tracing

YigitElma committed:
some clean up

YigitElma committed:
update docs

YigitElma committed:
update docs

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
update equinox version

YigitElma committed:
Merge branch 'rc/particles' into yge/particle-fit

f0uriest commented on desc/particles.py:
```diff
@@ -1242,3 +1281,195 @@ def _intfun_wrapper(
+            Equilibrium to be used to get transforms.
+
+        """
+        self.data_keys = ["B", "grad(|B|)", "e^rho", "e^theta", "e^zeta"]
```
`e^theta` diverges on axis so it can't be fit well, you can probably use `e^theta*sqrt(g)`

f0uriest replied:
also just to note, `e^rho` and `e^theta*sqrt(g)` don't scale like rho^m near axis so shouldn't be fit to a zernike basis

f0uriest reviewed (commented):
from an API standpoint, I'd prefer if you pass in the FourierChebyshevField as the field object rather than a dict of params, and pass the params separately the same way we do for an equilibrium (and allow the FourierChebyshevField to contain the params as an attribute)

YigitElma committed:
add ts to aux output to be able to use SaveAt(steps=True)

YigitElma committed:
update PID controller default coeffs, works better across solvers and particles

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'rc/particles' into yge/particle-fit

YigitElma committed:
add params_dict attribute

YigitElma committed:
use batched operations in fit and eval;uate

YigitElma committed:
add new quantity, make the procedure easier to implement for others setups

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'rc/particles' into yge/particle-fit

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'rc/particles' into yge/particle-fit

YigitElma committed:
add tests

YigitElma commented:
Not sure if this is necessary but we can reduce the number of fits by directly fitting the dot products. This will probably require higher fit resolution (since dot product increases the frequency window) but it may be more efficient. Currently, I fit 15 quantities, with the dot product you need 7 fits.

<details>
<summary>Here is the code implementation</summary>

```python
class VacuumGuidingCenterTrajectoryOpt(VacuumGuidingCenterTrajectory):
    @jit
    def vf(self, t, x, args):
        x = x.squeeze()
        model_args, field, params, kwargs = args
        m, q, mu = model_args
        xp, yp, zeta, vpar = x
        rho = jnp.sqrt(xp**2 + yp**2)
        theta = jnp.arctan2(yp, xp)
        # compute functions are not correct for very small rho
        rho = jnp.where(rho < 1e-6, 1e-6, rho)

        data = field.evaluate(rho, theta, zeta)

        # take dot product for rho, theta and zeta coordinates
        common = (m / q / data["|B|"] ** 2) * ((mu * data["|B|"] / m) + vpar**2)
        rhodot = common * data["(bxgrad(|B|))*e^rho"] # b . e^rho = 0
        thetadot_x_rho = vpar * data["b*e^theta*rho"] + common * data["(bxgrad(|B|))*e^theta*rho"]
        zetadot = vpar * data["b*e^zeta"] + common * data["(bxgrad(|B|))*e^zeta"]

        # get the derivative for cartesian-like coordinates
        xpdot = rhodot * jnp.cos(theta) - thetadot_x_rho * jnp.sin(theta)
        ypdot = rhodot * jnp.sin(theta) + thetadot_x_rho * jnp.cos(theta)
        # derivative the parallel velocity
        vpardot = -mu / m * data["b*grad(|B|)"]
        dxdt = jnp.array([xpdot, ypdot, zetadot, vpardot]).reshape(x.shape)
        return dxdt.squeeze()

class FourierChebyshevFieldOpt(IOAble):
    _static_attrs = ["L", "M", "N", "M_fft", "N_fft", "data_keys"]
    def __init__(self, L, M, N):
        self.L = L
        self.M = M
        self.N = N

    def build(self, eq):
        self.data_keys = ["b", "grad(|B|)", "e^rho", "e^theta*rho", "e^zeta", "|B|"]
        self.l = jnp.arange(self.L)
        self.M_fft = 2 * self.M + 1
        self.N_fft = 2 * self.N + 1
        self.m = jnp.fft.fftfreq(self.M_fft) * self.M_fft
        self.n = jnp.fft.fftfreq(self.N_fft) * self.N_fft
        x = jnp.cos(jnp.pi * (2 * self.l + 1) / (2 * self.L))
        rho = (x + 1) / 2
        self.grid = LinearGrid(rho=rho, M=self.M, N=self.N, sym=False, NFP=1)
        self.transforms = get_transforms(self.data_keys, eq, self.grid, jitable=True)

    def fit(self, params, profiles):
        data_raw = compute_fun(
            "desc.equilibrium.equilibrium.Equilibrium",
            self.data_keys,
            params,
            self.transforms,
            profiles,
        )
        L, M, N = self.L, self.M_fft, self.N_fft
        bxgb = jnp.cross(data_raw["b"], data_raw["grad(|B|)"])

        betr = dot(data_raw["b"], data_raw["e^theta*rho"])
        bez = dot(data_raw["b"], data_raw["e^zeta"])
        bxgbr = dot(bxgb, data_raw["e^rho"])
        bxgbtr = dot(bxgb, data_raw["e^theta*rho"])
        bxgbz = dot(bxgb, data_raw["e^zeta"])
        bdgb = dot(data_raw["b"], data_raw["grad(|B|)"])
        B = data_raw["|B|"]

        keys = ["betr", "bez", "bxgbr", "bxgbtr", "bxgbz", "bdgb", "B"]
        # stack data to perform 7 transforms in batch
        stacked_data = jnp.stack(
            [
                # ber.reshape(N, L, M),
                betr.reshape(N, L, M),
                bez.reshape(N, L, M),
                bxgbr.reshape(N, L, M),
                bxgbtr.reshape(N, L, M),
                bxgbz.reshape(N, L, M),
                bdgb.reshape(N, L, M),
                B.reshape(N, L, M),
            ]
        )
        coefs = jax.scipy.fft.dct(stacked_data, axis=2, norm=None)
        # handle the 0-th Chebyshev coefficient and normalization
        coefs = coefs.at[:, :, 0, :].divide(2)
        coefs /= self.L

        coefs = jnp.fft.fft(coefs, axis=3, norm=None)
        coefs = jnp.fft.fft(coefs, axis=1, norm=None)

        data = {}
        coefs_real = coefs.real
        coefs_imag = coefs.imag

        for i, key in enumerate(keys):
            data[key + "_real"] = coefs_real[i]
            data[key + "_imag"] = coefs_imag[i]

        data["l"] = self.l
        data["m"] = self.m
        data["n"] = self.n
        data["M"] = self.M_fft
        data["N"] = self.N_fft

        self.params_dict = data

    def evaluate(self, rho, theta, zeta, params=None):
        if params is None:
            params = self.params_dict

        # the cosine transforms reverses the order
        r0p = 1 - 2 * rho
        Tl = jnp.cos(params["l"] * jnp.arccos(r0p))
        m_theta = params["m"] * theta
        expm_real = jnp.cos(m_theta) / params["M"]
        expm_imag = jnp.sin(m_theta) / params["M"]

        n_zeta = params["n"] * zeta
        expn_real = jnp.cos(n_zeta) / params["N"]
        expn_imag = jnp.sin(n_zeta) / params["N"]

        # The new shape for these arrays will be (k, n, l, m) where k=15
        cf_real_all = jnp.stack(
            [
                params[key + "_real"]
                for key in [
                    "betr",
                    "bez",
                    "bxgbr",
                    "bxgbtr",
                    "bxgbz",
                    "bdgb",
                    "B",
                ]
            ]
        )
        cf_imag_all = jnp.stack(
            [
                params[key + "_imag"]
                for key in [
                    "betr",
                    "bez",
                    "bxgbr",
                    "bxgbtr",
                    "bxgbz",
                    "bdgb",
                    "B",
                ]
            ]
        )

        # "knlm,l->knm" contracts the 'l' dimension for all 'k' batches at once
        f_l_real = jnp.einsum("knlm,l->knm", cf_real_all, Tl)
        f_l_imag = jnp.einsum("knlm,l->knm", cf_imag_all, Tl)

        # "knm,m->kn" contracts the 'm' dimension for all 'k' batches
        f_lm_real = jnp.einsum("knm,m->kn", f_l_real, expm_real) - jnp.einsum(
            "knm,m->kn", f_l_imag, expm_imag
        )
        f_lm_imag = jnp.einsum("knm,m->kn", f_l_real, expm_imag) + jnp.einsum(
            "knm,m->kn", f_l_imag, expm_real
        )

        # "kn,n->k" contracts the 'n' dimension, leaving just the batch dimension
        results = jnp.einsum("kn,n->k", f_lm_real, expn_real) - jnp.einsum(
            "kn,n->k", f_lm_imag, expn_imag
        )

        out = {}

        out["b*e^theta*rho"] = results[0]
        out["b*e^zeta"] = results[1]
        out["(bxgrad(|B|))*e^rho"] = results[2]
        out["(bxgrad(|B|))*e^theta*rho"] = results[3]
        out["(bxgrad(|B|))*e^zeta"] = results[4]
        out["b*grad(|B|)"] = results[5]
        out["|B|"] = results[6]

        return out
```
</details>

YigitElma committed:
remove the error type just check the message, CI fails for weird reasons

YigitElma committed:
Merge branch 'master' into rc/particles

dpanici committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
Merge branch 'rc/particles' into yge/particle-fit

dependabot[bot] committed:
Update nbsphinx requirement from <=0.9.7,>=0.8.12 to >=0.8.12,<=0.9.8 (#2027)

Updates the requirements on
[nbsphinx](https://github.com/spatialaudio/nbsphinx) to permit the
latest version.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/spatialaudio/nbsphinx/releases">nbsphinx's
releases</a>.</em></p>
<blockquote>
<h2>nbsphinx 0.9.8</h2>
<p><a
href="https://pypi.org/project/nbsphinx/0.9.8/">https://pypi.org/project/nbsphinx/0.9.8/</a></p>
<ul>
<li>Re-enable Sphinx 8.2+</li>
<li>Support <code>text/x-rst</code> MIME type in raw cells</li>
<li>Support for <code>mathjax4_config</code></li>
</ul>
</blockquote>
</details>
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a
href="https://github.com/spatialaudio/nbsphinx/blob/master/NEWS.rst">nbsphinx's
changelog</a>.</em></p>
<blockquote>
<p>Version 0.9.8 -- 2025-11-28 -- PyPI__ -- diff__</p>
<ul>
<li>Re-enable Sphinx 8.2+</li>
<li>Support <code>text/x-rst</code> MIME type in raw cells</li>
<li>Support for <code>mathjax4_config</code></li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.8/">https://pypi.org/project/nbsphinx/0.9.8/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.7...0.9.8">https://github.com/spatialaudio/nbsphinx/compare/0.9.7...0.9.8</a></p>
<p>Version 0.9.7 -- 2025-03-03 -- PyPI__ -- diff__</p>
<ul>
<li>Disable Sphinx 8.2+ (for now)</li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.7/">https://pypi.org/project/nbsphinx/0.9.7/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.6...0.9.7">https://github.com/spatialaudio/nbsphinx/compare/0.9.6...0.9.7</a></p>
<p>Version 0.9.6 -- 2024-12-24 -- PyPI__ -- diff__</p>
<ul>
<li>Markdown: allow lists without leading blank line</li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.6/">https://pypi.org/project/nbsphinx/0.9.6/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.5...0.9.6">https://github.com/spatialaudio/nbsphinx/compare/0.9.5...0.9.6</a></p>
<p>Version 0.9.5 -- 2024-08-13 -- PyPI__ -- diff__</p>
<ul>
<li>Miscellaneous fixes</li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.5/">https://pypi.org/project/nbsphinx/0.9.5/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.4...0.9.5">https://github.com/spatialaudio/nbsphinx/compare/0.9.4...0.9.5</a></p>
<p>Version 0.9.4 -- 2024-05-06 -- PyPI__ -- diff__</p>
<ul>
<li>Require <code>docutils &gt;= 0.18.1</code></li>
<li>Minor fixes, documentation and CI updates</li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.4/">https://pypi.org/project/nbsphinx/0.9.4/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.3...0.9.4">https://github.com/spatialaudio/nbsphinx/compare/0.9.3...0.9.4</a></p>
<p>Version 0.9.3 -- 2023-08-27 -- PyPI__ -- diff__</p>
<ul>
<li>Fix gallery regression in Sphinx 7.2</li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.3/">https://pypi.org/project/nbsphinx/0.9.3/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.2...0.9.3">https://github.com/spatialaudio/nbsphinx/compare/0.9.2...0.9.3</a></p>
<p>Version 0.9.2 -- 2023-05-24 -- PyPI__ -- diff__</p>
<ul>
<li>Improve support for <code>sphinx_immaterial</code> theme</li>
<li>Improve support for links starting with <code>#</code></li>
<li>Add support for in-text citations</li>
<li>LaTeX: Add support for admonition titles</li>
</ul>
<p>__ <a
href="https://pypi.org/project/nbsphinx/0.9.2/">https://pypi.org/project/nbsphinx/0.9.2/</a>
__ <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.9.1...0.9.2">https://github.com/spatialaudio/nbsphinx/compare/0.9.1...0.9.2</a></p>
<p>Version 0.9.1 -- 2023-03-14 -- PyPI__ -- diff__</p>
<ul>
<li>pandoc: disable &quot;smart&quot; option only for version 2.0+</li>
</ul>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/4f07420639de321d63681ee8c560f73cb0c4a85f"><code>4f07420</code></a>
Release 0.9.8</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/75d6770733a5ef44553759449309a769c0329609"><code>75d6770</code></a>
Avoid a logging error</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/dd96f11edb14f7c2027c7ef50663f8c05a014278"><code>dd96f11</code></a>
DOC: Update URL for the &lt;img&gt; element</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/d97c0111c4af0423216d217234d87ed73013958d"><code>d97c011</code></a>
Add support for mathjax4_config (<a
href="https://redirect.github.com/spatialaudio/nbsphinx/issues/854">#854</a>)</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/4882668244e8c0d7b36d0743f9fbc1a48d3160da"><code>4882668</code></a>
Fix nbsphinx_assume_equations to work with newer versions of Sphinx</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/95372a74c543565b5e02357db5e248fb865a8326"><code>95372a7</code></a>
CI: use &quot;uv&quot; for &quot;publish&quot; workflow</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/9ce8c26cfa363a357b760ed1b0cc45ecb92de1f4"><code>9ce8c26</code></a>
setup.py -&gt; pyproject.toml</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/394b9ee5611109db6b234e44e2b61da8e939ac8d"><code>394b9ee</code></a>
DOC: use anonymous hyperlinks for embedded URLs</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/ab8592199d6637045b6f77992f10138590130d6b"><code>ab85921</code></a>
Bump actions/upload-artifact from 4 to 5</li>
<li><a
href="https://github.com/spatialaudio/nbsphinx/commit/76122c23c9568305eb6da15abd5278981e8960b1"><code>76122c2</code></a>
Bump actions/download-artifact from 4 to 6</li>
<li>Additional commits viewable in <a
href="https://github.com/spatialaudio/nbsphinx/compare/0.8.12...0.9.8">compare
view</a></li>
</ul>
</details>
<br />


Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Rory Conlin <wconlin@princeton.edu>

YigitElma committed:
minor updates

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
use proper vjp for proximal grad

YigitElma committed:
use proper vjp for proximal grad

YigitElma committed:
update changelog

YigitElma committed:
Merge branch 'master' into yge/proximal-grad

YigitElma committed:
remove the regularization

YigitElma committed:
add option to choose internal proximal inversion routine

YigitElma committed:
small simplification

YigitElma committed:
update the test, continuation causes bad approx. just optimize in single step

YigitElma committed:
Merge branch 'master' into yge/proximal-qr

YigitElma committed:
Merge branch 'master' into yge/proximal-qr

YigitElma committed:
Update desc/optimize/_constraint_wrappers.py

YigitElma committed:
Update desc/optimize/optimizer.py

YigitElma committed:
fix test

YigitElma committed:
Merge branch 'master' into rc/particles

YigitElma committed:
update changelog, add test

YigitElma committed:
update the notebook to use single step

YigitElma committed:
Merge branch 'yge/proximal-qr' into yge/particle-fit

YigitElma committed:
Merge branch 'rc/particles' into yge/particle-fit

YigitElma committed:
Merge branch 'master' into yge/particle-fit

YigitElma committed:
unmerge yge/proximal-qr

YigitElma committed:
Merge branch 'yge/proximal-grad' into yge/particle-fit

YigitElma committed:
revert qr stuff

YigitElma committed:
unmerge more

YigitElma committed:
update

YigitElma committed:
compute data on single NFP region, convert zeta to account the difference, update functions to accept FourierChebyshevField

YigitElma committed:
Merge branch 'master' into yge/particle-fit

YigitElma committed:
add e^theta*rho to unimplemented quantities

YigitElma committed:
update e^zeta computation, don't stack coefs multiple times, update tests

YigitElma committed:
slow but working spline version

YigitElma committed:
slightly more efficient

YigitElma committed:
Merge branch 'master' into yge/particle-fit

YigitElma committed:
add jit for profiling

YigitElma commented:
For some reason, this suffers from GPU synchronisation miserably, and for more particles, it is slower than direct approach.

YigitElma committed:
Merge branch 'master' into yge/particle-fit

YigitElma committed:
Merge branch 'master' into yge/particle-fit
